pkg/controller/controller_utils.go

```diff
@@ -542,6 +544,39 @@ func GetPodFromTemplate(template *v1.PodTemplateSpec, parentObject runtime.Objec
 	return pod, nil
 }
 
+func (r RealPodControl) validateNodeConflict(pod *v1.Pod) error {
+	if len(pod.Spec.NodeName) == 0 {
+		return nil
+	}
+
+	node, err := r.KubeClient.Core().Nodes().Get(pod.Spec.NodeName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	nodeInfo := schedulercache.NewNodeInfo()
+	err = nodeInfo.SetNode(node)
+	if err != nil {
+		return err
+	}
+
+	fit, reasons, err := predicates.PodSelectorMatches(pod, nil, nodeInfo)
```

Comment:
@k82cn I am not sure I understand. Is it the kubelet that decides that a Pod is not matching so it fails the Pod with MatchNodeSelector? If so, why can't the scheduler do this as well before it decides to assign a Pod on a Node?